File: src/lrs.cc
#include "lrs.h"

#include <cmath>
#include <fstream>
#include <iostream>

#include "score.h"

namespace colors {

inline int count_larger_than(const Vector& v, double value) {
  int count = 0;
  for (int i = 0; i < v.size(); ++i) {
    if (v[i] > value) ++count;
  }
  return count;
}

int calc_lrs(const double lambda, const Matrix& M, Matrix& L, Matrix& S) {
  int dim = M.rows();

  Matrix Y = M;  // copy
  L = Matrix::Zero(dim, dim);
  S = Matrix::Zero(dim, dim);
  Array Z = Array::Zero(dim, dim);

  Eigen::JacobiSVD<Matrix> svd_only_singlar_values(Y);
  const double norm_two = svd_only_singlar_values.singularValues()(0);
  const double norm_inf = Y.array().abs().maxCoeff() / lambda;
  const double dual_norm = std::max(norm_two, norm_inf);
  const double d_norm = M.norm();

  Y /= dual_norm;

  double mu = 1.25 / norm_two;
  const double rho = 1.5;
  const double mu_bar = mu * 1.0e+7;

  bool converged = false;
  int max_iter = 1000;
  double error_tolerance = 1.0e-7;
  int total_svd = 0;
  int sv = 10;
  int iter = 0;

  while (++iter < max_iter) {
    // update sparse matrix S
    Array temp_T = M - L + (1.0 / mu) * Y;
    S = (temp_T - lambda / mu).max(Z) + (temp_T + lambda / mu).min(Z);
    S = S.array().max(Z);

    // update low-rank matrix L
    Eigen::JacobiSVD<Matrix> svd(M - S + 1.0 / mu * Y,
                                 Eigen::ComputeFullU | Eigen::ComputeFullV);
    Matrix U = svd.matrixU();
    Matrix V = svd.matrixV();
    Vector singular_values = svd.singularValues();

    int svp = count_larger_than(singular_values, 1.0 / mu);
    if (svp < sv) {
      sv = std::min(svp + 1, dim);
    } else {
      sv = std::min(svp + static_cast<int>(0.05 * dim + 0.5), dim);
    }

    Matrix S_th =
        (singular_values.head(svp).array() - 1.0 / mu).matrix().asDiagonal();
    L = U.leftCols(svp) * S_th * V.leftCols(svp).transpose();
    L = L.array().max(Z);

    total_svd += 1;
    Matrix D = M - L - S;
    Y = Y + mu * D;

    // update mu
    mu = std::min(mu * rho, mu_bar);

    double objective = D.norm() / d_norm;
    if (objective < error_tolerance) {
      break;
    }
  }
  return 0;
}

int calc_lrs(const double lambda, const Matrix& M, const char* output_path) {
  Matrix L, S;
  calc_lrs(lambda, M, L, S);

  Matrix apc(M.cols(), M.cols());
  calc_apc(M, apc);

  save_lrs(M, S, apc, output_path);

  return 0;
}

int save_lrs(const Matrix& M, const Matrix& S, const Matrix& apc,
             const char* output_path) {
  std::ofstream fout;
  fout.open(output_path);
  if (fout.fail()) {
    std::cerr << "Failed in open file " << output_path << std::endl;
  }

  fout << "#pos_i pos_j original_score apc_score lrs_score\n";
  for (int i = 0; i < M.rows(); i++) {
    for (int j = i + 1; j < M.cols(); j++) {
      fout << i + 1 << " " << j + 1 << " " << M(i, j) << " " << apc(i, j) << " "
           << S(i, j) << std::endl;
    }
  }

  return 0;
}

}  // namespace colors

File: python/LRS.py
import sys
import numpy

def calc_lrs(lambda_, M):
    nrow, ncol = M.shape
    dim = np.minimum(nrow, ncol)
    Y = M
    
    L= np.zeros((nrow, ncol)); 
    S= np.zeros((nrow, ncol))
    Z= np.zeros((nrow, ncol))
    
    singular_values = np.linalg.svd(M, full_matrices=True, compute_uv=False)

    norm_two = np.max(singular_values)
    norm_inf = np.max(np.abs(Y)) / lambda_
    dual_norm = np.maximum(norm_two, norm_inf)
    
    d_norm = np.linalg.norm(M, ord=2)
    
    Y /= dual_norm
    
    mu = 1.25 / norm_two
    rho = 1.5
    mu_bar = mu * 1.0e+7
    converged = False
    max_iter = 10
    error_tolerance = 1.0e-7
    total_svd = 0
    sv = 10
    iter = 0
        
    for _ in range(max_iter):     
        
        #update sparse matrix S
        temp_T = M - L + (1.0 / mu) * Y;
        S = np.maximum(temp_T - lambda_ / mu, Z) + np.minimum(temp_T + lambda_ / mu, Z)
        S = np.maximum(S, Z)

        
        U, singular_values, V = np.linalg.svd(M - S + 1.0 / mu * Y, full_matrices=True,compute_uv=True)
        iter = iter+1

        svp = np.int(np.sum(np.greater(singular_values, 1.0 / mu)))
        print(svp, 'test')
        if svp < sv:
            sv = np.minimum(svp + 1, dim);
        else:
            sv = np.minimum(svp + np.int(0.05 * dim + 0.5), dim);

        S_th = np.diag(singular_values[:svp] - 1.0 / mu)

        L = np.matmul(np.matmul(U[:,:svp], S_th), V[:,:svp].transpose())
        L = np.maximum(L, Z)

        total_svd += 1;
        D = M - L - S;
        Y = Y + mu * D;

        mu = np.minimum(mu * rho, mu_bar);
        objective = np.linalg.norm(D, ord=2) / d_norm;
        print(objective, error_tolerance)
        if objective < error_tolerance:
            break;
      
    return S 
